### PR TITLE
add support for secondary_ip_addresses and finding eni by filters

### PR DESCRIPTION
**Pull request by kbroughton**
I found it a difficult to find eni_id as it is not returned by ec2_facts.
Other extras modules use the "filters" supported by boto to find ids by other properties,
so I employed that method.  Typically you would find eni_id by primary_ip_address.

Then I added support for secondary_ip_addresses which allows for virtual interfaces.
I chose to add this to ec2_eni as this module seems to manage the instance-interface-ips relationship.
It could also belong in ec2 module but that module is already pretty maxed out.

I have tested this with ansible 1.9, in a vpc on spot and regular 14.04 ubuntu instances.

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @kbroughton. @wimnat please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

**Comment by wimnat** [comment]
I like this.  I have actually been working on secondary_ips already so i'll compare to what is here.

The filter lookup is also a good idea.

Btw @kbroughton - easiest way to get eni_id is using the ec2_eni_facts module.

File: cloud/amazon/ec2_eni.py [hunk 1]
```diff
@@ -27,9 +27,14 @@
       - The ID of the ENI
     required: false
     default: null
+  filters:
+    description:
+      - Dictionary of key-value pairs that can be searched to locate eni_id if it is not known.
```
Comment (wimnat) [inline cloud/amazon/ec2_eni.py:32]: I would change to "... to locate eni if eni id is not known"

File: cloud/amazon/ec2_eni.py [hunk 2]
```diff
@@ -109,6 +138,20 @@
     description: "My new description"
     state: present
 
+# Identify an insctance id by primary-ip-address and instance-id, then modify
```
Comment (wimnat) [inline cloud/amazon/ec2_eni.py:142]: Typo on instance

**Comment by gregdek** [comment]
Thanks @kbroughton for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

**Comment by gregdek** [comment]
Thanks @kbroughton for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by kbroughton** [comment]
I incorporated the sorted(security_groups) which looked like the only change causing conflict.

**Comment by wimnat** [comment]
@kbroughton - secondary ip addresses are now supported thanks to a recent merge.  Maybe close this and create a new PR to add the lookup by filters functionality?

needs_revision

**Comment by brandond** [comment]
This module desperately needs the ability to detect if change is required; as it currently stands this module is not truly idempotent - it tries to create and attach a new ENI even if a suitable one is already attached to the selected instance.

**Comment by wimnat** [comment]
@brandond can you define 'suitable one'?

**Comment by brandond** [comment]
@wimnat I would probably go with:
- In the correct subnet
- Attached to the correct instance
- At the proper index on the instance

If a given ENI matches those criteria, a new one should not be created and attached but rather reused; any secondary IPs (and tags; ahem) should be updated.

I agree with you that this PR should probably be closed though; https://github.com/ansible/ansible-modules-extras/pull/2365 seems to be more up-to-date.

**Comment by wimnat** [comment]
close_me closeme

**Comment by gregdek** [comment]
Thanks @kbroughton. One of the current maintainers, @wimnat requested to close this PR.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@kbroughton A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**